diff --git a/net/bridge/br_input.c b/net/bridge/br_input.c
index 6e191d34..003c586f 100644
--- a/net/bridge/br_input.c
+++ b/net/bridge/br_input.c
@@ -41,6 +41,7 @@ extern void br_hrdb_update(const unsigned char *mac, unsigned char event_sent);
 
 #define AUTHSRV_PKT 3001
 #define WIFIDOG_PKT 2060 /* FIXME */
+#define WIFIDOG_PKT1 2061 /* FIXME */
 #define HTTP_PKT    80
 #define HTTPS_PKT   443
 #define DNS_PKT     53
@@ -270,6 +271,7 @@ int br_handle_frame_finish(struct sk_buff *skb)
 			 * the mark rule execute even before the dst
 			 * mac on the HTTP(S) packets are replaced. */
 			case WIFIDOG_PKT:
+			case WIFIDOG_PKT1:
 				//if (memcmp((unsigned char *)dest, br->dev->dev_addr, ETH_ALEN) == 0) {
 				//	break; /* As we are using bridge-nf-call-iptables commenting this*/
 				//} else if (skb->mark != 0x3) {
