diff --git a/net/core/skbuff_recycle.c b/net/core/skbuff_recycle.c
index 32be06f..1467838 100644
--- a/net/core/skbuff_recycle.c
+++ b/net/core/skbuff_recycle.c
@@ -102,6 +102,7 @@ inline struct sk_buff *skb_recycler_alloc(struct net_device *dev, unsigned int l
 
 	if (likely(skb)) {
 		struct skb_shared_info *shinfo;
+        bool pfmemalloc = skb->pfmemalloc;
 
 		/* We're about to write a large amount to the skb to
 		 * zero most of the structure so prefetch the start
@@ -113,6 +114,7 @@ inline struct sk_buff *skb_recycler_alloc(struct net_device *dev, unsigned int l
 
 		zero_struct(skb, offsetof(struct sk_buff, tail));
 		atomic_set(&skb->users, 1);
+        skb->pfmemalloc = pfmemalloc;
 		skb->mac_header = (typeof(skb->mac_header))~0U;
 		skb->transport_header = (typeof(skb->transport_header))~0U;
 		zero_struct(shinfo, offsetof(struct skb_shared_info, dataref));
