diff --git a/net/bridge/br_input.c b/net/bridge/br_input.c
index 1973ca3..5b7d4fc 100644
--- a/net/bridge/br_input.c
+++ b/net/bridge/br_input.c
@@ -54,7 +54,7 @@ extern void br_hrdb_update(const unsigned char *mac, unsigned char event_sent);
 #define NETBIOS_PKT 137
 #define LKLOCAL_PKT 5355
 #define INVALID_PKT 0xFFFF
-
+#define L2PROT_PKT  0x88B7
 struct net_device *br_tuntap_dev;
 struct net_device *url_tuntap_dev;
 extern int url_flag;
@@ -241,7 +241,10 @@ int br_handle_frame_finish(struct sk_buff *skb)
 	if (!br_allowed_ingress(p->br, nbp_get_vlan_info(p), skb, &vid))
 		goto out;
 
-	/* insert into forwarding database after filtering to avoid spoofing */
+    if (eth_hdr(skb)->h_proto == htons(L2PROT_PKT)) {
+        goto drop;
+    }
+    /* insert into forwarding database after filtering to avoid spoofing */
 	br = p->br;
 	if (p->flags & BR_LEARNING)
 		br_fdb_update(br, p, eth_hdr(skb)->h_source, vid, false);
